lib/pacemaker/pcmk_cluster_queries.c

```diff
@@ -228,6 +230,47 @@ designated_controller_event_cb(pcmk_ipc_api_t *controld_api,
     data->rc = pcmk_rc_ok;
 }
 
+static void
+node_name_event_cb(pcmk_ipc_api_t *controld_api, enum pcmk_ipc_event event_type,
+                   crm_exit_t status, void *event_data, void *user_data)
+{
+    data_t *data = (data_t *) user_data;
+    pcmk__output_t *out = data->out;
+    const pcmk_controld_api_reply_t *reply = NULL;
+
+    int rc = validate_controld_reply(data, controld_api, event_type, status,
+                                     event_data, pcmk_controld_reply_info);
+
+    if (rc != pcmk_rc_ok) {
+        return;
+    }
+
+    reply = (const pcmk_controld_api_reply_t *) event_data;
+
+    if (reply->data.node_info.uname == NULL) {
+        out->err(out, "Node is not known to cluster");
+        data->rc = pcmk_rc_node_unknown;
+        return;
+    }
+
+    if (data->show_output) {
+        /* Using data->node_id outputs "local node" for some formats if
+         * pcmk__node_name() is called with node_id=0. Using the reply's node ID
+         * would show the true cluster node ID. Neither option is clearly better
+         * than the other.
+         */
+        out->message(out, "node-name",
+                     *(data->node_id), reply->data.node_info.uname);
+    }
+
+    if (data->node_name != NULL) {
+        *(data->node_name) = strdup(reply->data.node_info.uname);
+        CRM_ASSERT(*(data->node_name) != NULL);
+    }
```

Comment:
see pcmk__str_update()

Reply:
I did that before this; it segfaulted if the pointer was uninitialized, e.g.,
```
// It actually took two calls to segfault
char *node_name1;
char *node_name2;
pcmk__node_name(..., &node_name1, ...);
pcmk__node_name(..., &node_name2, ...);
```

We can get around that by explicitly nulling the arg at the beginning of `pcmk__node_name()`:
```
pcmk__node_name(...)
{
    if (node_name != NULL) {
        *node_name = NULL;
    }
}
```
But that's not any shorter overall. I have no preference. I'll probably change it to do the explicit nulling.

Reply:
That seems reasonable